rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala

```diff
@@ -0,0 +1,271 @@
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.ncf
+
+import net.liftweb.common.Loggable
+import java.nio.file.Paths
+import java.nio.file.Files
+import scala.xml.NodeSeq
+import java.nio.charset.StandardCharsets
+import com.normation.inventory.domain.AgentType
+import net.liftweb.common.Box
+import com.normation.rudder.repository.xml.GitArchiverUtils
+import com.normation.cfclerk.services.GitRepositoryProvider
+import java.io.File
+import com.normation.rudder.repository.xml.RudderPrettyPrinter
+import com.normation.rudder.repository.GitModificationRepository
+import com.normation.rudder.repository.xml.GitArchiverFullCommitUtils
+import com.normation.eventlog.ModificationId
+import org.eclipse.jgit.lib.PersonIdent
+import com.normation.rudder.repository.GitPath
+import net.liftweb.common.Full
+import com.normation.eventlog.EventActor
+import com.normation.rudder.services.user.PersonIdentService
+
+class TechniqueWriter(
+  archiver : TechniqueArchiver
+) extends Loggable {
+
+  private[this] var agentSpecific = new ClassicTechniqueWriter :: new DSCTechniqueWriter :: Nil
+
+  def commonMetadata(technique : Technique, methods: Map[BundleName, GenericMethod]) = {
+
+    <TECHNIQUE name={technique.name}>
+      <DESCRIPTION>technique.description</DESCRIPTION>
+      {for {
+        techWriter <- agentSpecific
+      } yield {
+        techWriter.techniqueMetadata(technique, methods)
+      }}
+      <SECTIONS> {
+        for {
+          (name, methodCalls) <- technique.methodCalls.groupBy(_.methodId).toList.sortBy(_._1.value)
+          if ! name.value.startsWith("_")
+        } yield {
+          <SECTION component="true" multivalued="true" name={name.value}>
+            <REPORTKEYS>
+              { for {
+                  call <- methodCalls
+                  method <- methods.get(call.methodId)
+                } yield {
+                  val value = call.parameters.get(method.classParameter).getOrElse("")
+                  <VALUE>{value}</VALUE>
+                }
+              }
+            </REPORTKEYS>
+          </SECTION>
+
+        }
+      }
+      </SECTIONS>
+    </TECHNIQUE>
+  }
+
+  def writeAll(technique : Technique, methods: Map[BundleName, GenericMethod], modId : ModificationId, commiter : EventActor) : Box[Unit] = {
```

Comment:
I believe we would gain a lot here with a dedicated error type and custom cases, ie:
sealed trait MetadataGenError extends Exception  //still wondering if extending exception is the correct thing to do, but we need at least a string description and most of the time, it is handy to be able to capture a stack trace from elsewhere)
case object IOError(msg: String, ex: Throwable) extends MetadataGenError ...

writeTechnique(...): Either[MetadataGenError, Unit] = { ... }

And we will handle toBox at the API level. 

It will be more understandable for a reader (and even in logs), and we won't as easely miss cases where the reason of the error get lost. For ex, if an exception happen in writeTechnique below, I think it can quite hard to know what caused the "writeAll" error. If it's a "WriteTechniqueIOError(msg)", quite obvious.

Reply:
Not done yet, in next commit